### PR TITLE
DefaultHeaders#valueIterator to support removal

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
While iterating values it is often desirable to be able to remove individual
entries. The existing mechanism to do this involves removal of all entries and
conditional re-insertion which is heavy weight in order to remove a single
value.

Modifications:
- DefaultHeaders$ValueIterator supports removal

Result:
It is possible to remove entries while iterating the values in DefaultHeaders.

### CODE REVIEW

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 1]
```diff
@@ -1073,14 +1083,18 @@ public V next() {
...
             calculateNext(next.next);
-            return current.value;
+            return previous.value;
         }
 
         @Override
         public void remove() {
-            throw new UnsupportedOperationException("read only");
+            if (previous == null) {
+                throw new IllegalStateException();
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: I am not sure how to feel about this honestly as it is not really the API contract of an `Iterator`.

Comment (Scottmitch) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: IIUC this is required by the API contract: 

```java
     * @throws IllegalStateException if the {@code next} method has not
     *         yet been called, or the {@code remove} method has already
     *         been called after the last call to the {@code next}
     *         method
```

Comment (normanmaurer) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: TIL.... never noticed this before in the `Iterator` javadocs.

Comment (normanmaurer) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: As this is required can you also add a test case for it ?

Comment (Scottmitch) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: done

File: codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java [hunk 2]
```diff
@@ -1012,6 +1012,15 @@ private V remove0(int h, int i, K name) {
...
     }
 
+    private void remove0(HeaderEntry<K, V> entry) {
+        int i = index(entry.hash);
+        HeaderEntry<K, V> e = entries[i];
+        if (e == entry) {
+            entries[i] = entry.next;
+        }
+        entry.remove();
+    }
```
Comment (normanmaurer) [inline codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: I think you also need to call `--size;`

Can you please also adjust the unit test to check for this ?

Comment (Scottmitch) [reply codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java]: good catch! fixed.

**Comment by normanmaurer** [comment]
@Scottmitch thanks a lot...

**Pull request merged**